### ISSUE TITLE
"none" driver should bind to the localhost interface by default

### ISSUE THREAD

**Issue by tstromberg** [kubernetes/minikube#4313]
For improved security by default.

**Comment by tstromberg** [comment]
Related: #2762

**Comment by vaibhavk** [comment]
I would like to work on this issue

**Comment by tstromberg** [comment]
@vaibhavk - Yes, I would love help on this. Much of this goes toward rectifying https://github.com/kubernetes/minikube/blob/master/docs/vmdriver-none.md#decreased-security - but in the future with Docker/Podman deployments, it will also be necessary.

What I propose is that when run with the `none` driver, the services listed in that document should all bind to 127.0.0.1 by default. We'll also use this in the future when we support docker/podman deployments. Some further implementation details:

* apiserver supports `--bind-address`.  You can add flags to the apiserver config in https://github.com/kubernetes/minikube/blob/master/pkg/minikube/bootstrapper/kubeadm/templates.go
 -- see how we already add AdvertiseAddress.

* kubelet supports --address. In that same file, there is kubeletSystemdTemplate which generates the right flags.

One approach for kubelet might be to have code that adds ExtraOptions if the driver is none. See https://github.com/kubernetes/minikube/blob/33c217eedcb93357df18cd7b2534014c46866d37/pkg/minikube/bootstrapper/kubeadm/kubeadm.go#L430

Let me know if you would like more guidance. Feel free to reach out on Slack #minikube as well if you prefer real-time discussion. Thank you!

**Comment by medyagh** [comment]
@vaibhavk are you still working on this ?

**Comment by vaibhavk** [comment]
@medyagh Yes, I'm back on it

**Comment by afbjorklund** [comment]
If we bind everything to localhost on the VM, how will you access it from the developer machine ?

**Comment by afbjorklund** [comment]
This feature would be another reason why generic (#4733) is needed. The current workaround of running --vm-driver=none on a _remote_ VM would no longer work properly after this, if it only listens on localhost. It would require you to ssh into the control plane from your developer machine, in order to reach the apiserver. It's probably a good feature for `none`, though. It was never supposed to expose it outside of localhost.

**Comment by itsallonetome** [comment]
I've been struggling with this ... an alternative (which would work for me) is an option to specify whether minikube should take its external interface to be the host's IP address (as at present) or localhost or the docker bip gateway.

At present, I can find no way to force minikube to use localhost.  using `--extra-config kubelet.node-ip="127.0.0.1"` breaks multi-pod systems as the pods listen on 127.0.0.1 but other pods try to talk to them on the hosts's external IP address.

I get:
`$ kubectl cluster-info
Kubernetes master is running at https://localhost:8443

$ minikube ip
10.74.54.212`

I can't use the other drivers as the VMs in which I'm working have virtualisation disabled.

**Comment by avisiedo** [comment]
Hi.... I have a situation where I would like just the opposite. I mean, I am starting the cluster with minikube, and I would like the sockets to be bound to 0.0.0.0 so that I can access from another host. I now that minikube is not for production, and it is not my intention to use this for production, but make this bind to external IP is good for me. I am using --vm-driver=none to start the kubernetes cluster.

So in short: I could I start minikube to allow it to bind to 0.0.0.0?? Any configuration file to set up to allow this?? I am just starting with Kubernetes, sorry if I don't have enough skills yet.

Thanks!

**Comment by ykfq** [comment]
Still need this feature when there are mutilple nics.

**Comment by medyagh** [comment]
@vaibhavk  are you still interested to do this ?

**Comment by vaibhavk** [comment]
@medyagh I recently switched to Darwin & not able to test the build, is there any work around for this? I'll be happy to work on this again and send PR

**Comment by medyagh** [comment]
> Hi.... I have a situation where I would like just the opposite. I mean, I am starting the cluster with minikube, and I would like the sockets to be bound to 0.0.0.0 so that I can access from another host. I now that minikube is not for production, and it is not my intention to use this for production, but make this bind to external IP is good for me. I am using --vm-driver=none to start the kubernetes cluster.
> 
> So in short: I could I start minikube to allow it to bind to 0.0.0.0?? Any configuration file to set up to allow this?? I am just starting with Kubernetes, sorry if I don't have enough skills yet.
> 
> Thanks!

@avisiedo opens ups big security problems,since minikbue is directed at develoipers for local kubernetes, that would be a bad default ! however I would accept any PR that would add that as an optional feature. (with extra warnning to the user that if they take the risk we could allow that)

**Comment by medyagh** [comment]
> @medyagh I recently switched to Darwin & not able to test the build, is there any work around for this? I'll be happy to work on this again and send PR
@vnzongzna 
yes ofcourse you can still work on this. please let me know if you need any help on the PR review

**Comment by KubaJakubowski** [comment]
> Hi.... I have a situation where I would like just the opposite. I mean, I am starting the cluster with minikube, and I would like the sockets to be bound to 0.0.0.0 so that I can access from another host. I now that minikube is not for production, and it is not my intention to use this for production, but make this bind to external IP is good for me. I am using --vm-driver=none to start the kubernetes cluster.
> 
> So in short: I could I start minikube to allow it to bind to 0.0.0.0?? Any configuration file to set up to allow this?? I am just starting with Kubernetes, sorry if I don't have enough skills yet.
> 
> Thanks!

Anything new on this? 
I'm looking for way to access minikube pod in cluster within local network and I cannot do it in any way except using socat.

In case that someone would like to use it anyway with socat I post solution below:
`socat TCP-LISTEN:<HOST_PORT>,fork TCP:<CLUSTER_ADDRESS>:<POD_PORT> &`

**Comment by tstromberg** [comment]
@KubaJakubowski - That would be a completely different issue from binding to localhost, and well out of the scope of minikube. 

My personal recommendation would be an SSH tunnel, but socat would probably work as well. Alternatively, you should probably just run `kubeadm` directly if you need network access. It's what minikube runs underneath.
